File: artifacts/api-server/py/routes/portfolio.py
from fastapi import APIRouter, HTTPException
from schemas import PortfolioOverviewOut, Position, Allocation
from auth_utils import CurrentUser
import alpaca_client

router = APIRouter()


@router.get("/portfolio/overview", response_model=PortfolioOverviewOut)
def portfolio_overview(user: CurrentUser):
    if not alpaca_client.is_configured():
        raise HTTPException(status_code=503, detail="Alpaca not configured")

    try:
        account = alpaca_client.get_account()
        raw_positions = alpaca_client.get_positions()

        equity = float(account.get("equity", 0))
        last_equity = float(account.get("last_equity", equity))
        total_pnl = equity - last_equity
        total_pnl_pct = ((equity - last_equity) / last_equity * 100) if last_equity else 0

        positions = []
        for p in raw_positions:
            side = "long" if float(p.get("qty", 0)) > 0 else "short"
            qty = abs(int(float(p.get("qty", 0))))
            entry = float(p.get("avg_entry_price", 0))
            current = float(p.get("current_price", 0))
            unrealized = float(p.get("unrealized_pl", 0))
            positions.append(Position(
                symbol=p.get("symbol", ""),
                side=side,
                quantity=qty,
                entryPrice=entry,
                currentPrice=current,
                unrealizedPnL=unrealized,
                broker="Alpaca",
            ))

        return PortfolioOverviewOut(
            totalBalance=equity,
            totalPnL=total_pnl,
            totalPnLPercent=round(total_pnl_pct, 2),
            openPositions=len(positions),
            allocations=[
                Allocation(broker="Alpaca", balance=equity, percentage=100.0)
            ],
            positions=positions,
        )
    except Exception as e:
        raise HTTPException(status_code=502, detail=f"Alpaca error: {str(e)}")

File: artifacts/api-server/src/routes/portfolio.ts
import { Router } from "express";
import { requireAuth } from "../middleware/requireAuth";

const router = Router();

router.get("/portfolio/overview", requireAuth, (_req, res) => {
  res.json({
    totalBalance: 71200.00,
    totalPnL: 21200.00,
    totalPnLPercent: 42.4,
    openPositions: 2,
    allocations: [
      { broker: "Alpaca", balance: 71200.00, percentage: 100 },
    ],
    positions: [
      {
        symbol: "AAPL",
        side: "long",
        quantity: 75,
        entryPrice: 228.50,
        currentPrice: 231.20,
        unrealizedPnL: 202.50,
        broker: "Alpaca",
      },
      {
        symbol: "TSLA",
        side: "long",
        quantity: 50,
        entryPrice: 250.30,
        currentPrice: 248.10,
        unrealizedPnL: -110.00,
        broker: "Alpaca",
      },
    ],
  });
});

export default router;
